py/asmrv32.c

```diff
@@ -445,12 +458,14 @@ void asm_rv32_emit_mov_reg_local(asm_rv32_t *state, mp_uint_t rd, mp_uint_t loca
     mp_uint_t lower = 0;
     split_immediate(offset, &upper, &lower);
 
-    // lui   rd, HI(offset) ; Or c.lui if possible
-    // c.add rd, sp
-    // lw    rd, LO(offset)(rd)
-    load_upper_immediate(state, rd, upper);
-    asm_rv32_opcode_cadd(state, rd, ASM_RV32_REG_SP);
-    asm_rv32_opcode_lw(state, rd, rd, lower);
+    // TODO: Can this clobber REG_TEMP[0:2]?
+
+    // lui   temporary, HI(offset) ; Or c.lui if possible
+    // c.add temporary, sp
+    // lw    rd, LO(offset)(temporary)
+    load_upper_immediate(state, INTERNAL_TEMPORARY, upper);
```

Comment:
Why did this change to use `INTERNAL_TEMPORARY` instead of `rd`?  It seems like using `rd` is OK and then it doesn't require an additional register.

Reply:
Too eager when copying and pasting I guess.  `INTERNAL_TEMPORARY` is always saved and restored as part of the function prologue/epilogue and its clobbering is always function-scoped so it's safe.  I'll fix that to use `rd` though, thanks.

Reply:
There are other places where the temporary register is used, like in `asm_rv32_emit_jump` and `asm_rv32_emit_call_ind` for example.

If I have to pick another register to replace that with, I'd say I could use REG_TEMP2 instead.  I understand that right now is only used inside `emit_native_rot_three` and in viper load/stores for RV32, but I wonder if there are any chances of that register being used elsewhere in the future.  If that's OK for me to use then this may allow few more size optimisations, as `S0` is inside the compressed opcodes register window.

Reply:
If using `INTERNAL_TEMPORARY` instead of `rd` allows for more size optimisations, then that's a good reason to use `INTERNAL_TEMPORARY`.

Reply:
I've looked a bit more into this to see if by shuffling things around there are more avenues for size optimisation.

Something that can be done (not in this PR, of course), would be to free up the register taken by `INTERNAL_TEMPORARY` (`S0`) and map it to `REG_TEMP0`.  Whilst this will still require a push/pop of the register in question (S-registers have to be saved and restored in this case), it would allow usage of compressed opcodes like `C.SUB`, `C.AND`, `C.OR` and `C.XOR`, and make it viable to use `C.SW` in more places.  There would be a penalty when calling out to the MicroPython runtime as REG_TEMP0 would probably be already clobbered when calling `asm_rv32_emit_call_ind`.

I'd need to come up with some numbers to see if such a thing is justified, though.

Anyway, I'll change the code to use `rd` and `REG_TEMP2` whenever it won't impact on code size.  Hopefully there's a way to make this bit of code better than it is now.